docs/core/v3.0.rst

```diff
@@ -0,0 +1,1761 @@
+
+    The chunk shape elements are non-zero when the corresponding dimensions of
+    the arrays are of non-zero length.
+
+.. _grid:
+.. _grids:
+
+*Grid*
+
+    The chunks_ of an array_ are organised into a grid. This
+    specification only considers the case where all chunks_ have the
+    same chunk shape and the chunks form a regular grid. However,
+    extensions may define other grid types such as
+    rectilinear grids.
+
+.. _memory layout:
+.. _memory layouts:
+
+*Memory layout*
+
+    An array_ is associated with a memory layout which defines how to
+    construct a binary representation of a single chunk_ by organising
+    the binary values of the elements_ within the chunk_ into a single
+    contiguous sequence of bytes. This specification defines two types
+    of memory layout based on "C" (row-major) and "F" (column-major)
+    ordering of elements_, but extensions may define other
+    memory layouts.
+
+.. _compressor:
+.. _compressors:
+
+*Compressor*
+
+    An array_ may be associated with a compressor, which is a codec_
+    that transforms the binary representation of a chunk_ in some way,
+    usually to reduce data size.
+
+.. _codec:
+.. _codecs:
+
+*Codec*
+
+    A codec is a pair of algorithms which transform binary data in
+    some way and are used to encode and decode chunks_. This
+    specification defines a codec interface which comprises
+    a pair of operations, one to perform the transformation (encode),
+    the other to reverse the transformation (decode). This
+    specification only considers the case where a codec is used as a
+    compressor_, but extensions may extend the chunk encoding
+    process, for example, to add support for one or more filter codecs
+    to be applied prior to compression.
+
+.. _metadata document:
+.. _metadata documents:
+
+*Metadata document*
+
+    Each array_ in a hierarchy_ is represented by a metadata document,
+    which is a machine-readable document containing essential
+    processing information about the node. For example, an array_
+    metadata document will specify the number of dimensions_, shape_,
+    `data type`_, grid_, `memory layout`_ and compressor_ for that
+    array_.
+
+    Groups can have a optional metadata document which provide extra
+    information about a group.
+
+.. _store:
+.. _stores:
+
+*Store*
+
+    The `metadata documents`_ and encoded chunk_ data for all nodes in
+    a hierarchy_ are held in a store. To enable a variety of different
+    store types to be used, this specification defines an `Abstract store
+    interface`_ which is a common set of operations that stores may
+    provide.
+
+.. _storage transformer:
+.. _storage transformers:
+
+*Storage transformer*
+
+    To enhance the storage capabilities, storage transformers may
+    change the storage structure and behaviour of data coming from
+    an array_ in the underlying store_. Upon retrieval the original data
```

Comment:
Here a "chunk storage transformer" would be a special case of a more generic storage transformer. You could also have a "metadata transformer" or a transformer for any of the store features. A metadata transformer would be sufficient to enable consolidated metadata.

Reply:
Sounds good IMO :+1: For array-specific transformers, such as sharding, it makes sense to allow array-specific configurations, e.g. having different shard sizes per array. Having `global storage transformers` and `chunk storage transformers` seems to be a good option. One limitation of global storage transformers is that the top-level metadata must still be in place to allow the setup of the zarr dataset, but other metadata might still be transformed (e.g. having lookups in the global/root metadata for the consolidated metadata transformer)

Reply:
Yes, it wouldn't make sense to specify sharding at the root. I think we could also distinguish between a feature having a possible implementation in terms of some sort of storage transformation, and how it is specified in the metadata.

To me it seems most logical to include within a list of storage transformers in the metadata things that are fully or mostly compostable with other storage transformers in a generic way, i.e. handles arbitrary key/value pairs.  For example, something that does content-addressable storage, or stores everything in an sqlite or lmdb database could make sense as a storage transformer.

Sharding could be implemented as a storage transformation, but it only makes sense as the first (innermost) transformation, as it relies on knowledge of the chunk keys.  Additionally, it is useful to expose the shard size in the API so that users know what granularity should be used for efficient writes.  Therefore in my mind it would be more logical to just make sharding its own extension rather than try to put it in a generic storage transformers field.  Implementations would still be free to implement it as a storage transformation.

Similarly, consolidated metadata could be implemented as a storage transformation, but as it presumably just store everything in the root metadata document, it would not be meaningfully composable with anything else.  Therefore I think it would also be more logical as just its own extension.

Reply:
The idea of global storage transformers is kept in #169, and @rabernat wants to provide a PR for it.

The order or storage transformers is not flexible and needs special care, but I think that exactly this problem is solved (at least for the spec, not implementations) by having a stack of storage transformers. :+1: for having recommendations about the orders of the transformers for implementations.